src/java/org/apache/cassandra/repair/RepairJob.java

```diff
@@ -189,30 +188,33 @@ public void onFailure(Throwable t)
                     });
                 });
             }
-
-            // When all snapshot complete, send validation requests
-            treeResponses = allSnapshotTasks.flatMap(endpoints -> {
-                if (parallelismDegree == RepairParallelism.SEQUENTIAL)
-                    return sendSequentialValidationRequest(allEndpoints);
-                else
-                    return sendDCAwareValidationRequest(allEndpoints);
-                }, taskExecutor);
         }
         else
         {
-            // If not sequential, just send validation request to all replica
-            treeResponses = paxosRepair.flatMap(input -> sendValidationRequest(allEndpoints));
+            allSnapshotTasks = null;
         }
-        treeResponses = treeResponses.map(a -> {
-            state.phase.validationCompleted();
-            return a;
-        });
+
+        // Run validations and the creation of sync tasks in the scheduler, so it can limit the number of Merkle trees
+        Scheduler.Task<List<SyncTask>> syncTasks = new Scheduler.Task<>()
+        {
+            @Override
+            public void run()
+            {
+                createSyncTasks(paxosRepair, allSnapshotTasks, allEndpoints).addCallback((s, f) -> {
+                    if (f != null)
+                        tryFailure(f);
+                    else
+                        trySuccess(s);
+                });
+            }
+        };
+        session.validationScheduler.schedule(session.getId(), taskExecutor, syncTasks);
```

Comment:
feel its cleaner to return `Future<T>` from `schedule`

Reply:
Done.